asset-transfer-basic/chaincode-java/README.md

```diff
@@ -0,0 +1,164 @@
+
+## Basic asset transfer 
+
+This project demonstrates the use of the Java SDK, running a basic asset transfer contract using the "chaincode as a service" 
+pattern.
+
+
+## Setup 
+
+In this sample we will employ the [Kubernetes Test Network](../../test-network-k8s) to illustrate a scenario of 
+building, running, and debugging chaincode on a development workstation.  
+
+This project is also compatible with the legacy chaincode builder pipeline and the compose based test-network.
+For additional details, see the [End-to-end with the test-network](../../test-network/CHAINCODE_AS_A_SERVICE_TUTORIAL.md#end-to-end-with-the-the-test-network)
+documentation.
+
+## [Quickstart](../../test-network-k8s#quickstart)
+
+```shell
+export PATH=${PWD}/../../test-network-k8s:$PATH 
+
+network kind 
+
+network up 
+network channel create
+```
+
+```shell
+network chaincode deploy  ${PWD}
+
+network chaincode invoke  asset-transfer-basic '{"Args":["InitLedger"]}'
+network chaincode query   asset-transfer-basic '{"Args":["ReadAsset","asset1"]}' | jq 
+```
+
+## Detailed Guide 
+
+```shell
+network down
+network up
+network channel create
+```
+
+```shell
+# Build the chaincode docker image 
+docker build -t hyperledger/fabric-samples/asset-transfer-basic/chaincode-java .
+
+# Load the docker image directly to the KIND control plane.  
+# (Alternately, build/tag/push the image to a remote container registry, e.g. localhost:5000) 
+kind load docker-image hyperledger/fabric-samples/asset-transfer-basic/chaincode-java
+```
+
+```shell
+# Assemble the chaincode package archive 
+network chaincode package $PWD/ccpackage/ $PWD/build/asset-transfer.tgz
+
+# Determine the ID for the chaincode package 
+CORE_CHAINCODE_ID_NAME=$(network chaincode id $PWD/build/asset-transfer.tgz)
+
+# Launch the chaincode in k8s as Deployment + Service 
+network chaincode launch $PWD/build/asset-transfer.tgz
+
+# Complete the chaincode lifecycle 
+network chaincode install $PWD/build/asset-transfer.tgz 
+network chaincode approve asset-transfer-basic $CORE_CHAINCODE_ID_NAME
+network chaincode commit  asset-transfer-basic
+```
+
+```shell
+# execute the smart contract by name 
+network chaincode invoke  asset-transfer-basic '{"Args":["InitLedger"]}'
+network chaincode query   asset-transfer-basic '{"Args":["ReadAsset","asset1"]}'
+```
+
+```shell
+kubectl -n test-network logs -f deployment/org1peer1-cc-asset-transfer-basic
+```
+
+## Debugging 
+
+### Build
+
+```shell
+./gradlew shadowJar
```

Comment:
For debug, you can either launch from the IDE or attach to a running process from the IDE.

Matter of preference to an extent. 

Creating the shadow jar might be irrelevant - and really only needed for production deployment.  Plus if you are Eclipse say, then you can just hit run! 

Or maybe just later, we do `./gradlew run` so that compiles and starts? then attach?

Reply:
`gradle run` would be a nice addition here.  IMO it should be managed in the context of a build / gradle plugin.

The trick to the run target is in picking up and determining the CC ID, which need to be in the scope of main()'s env.  This is straightforward, especially if the gradle task is constructing the chaincode package archive.    Note that this does not _need_ to be in a gradle plugin - it should be possible to shell out to generate the sha256 hash, assemble the package, etc. from a gradle build script.

a `run` target would be nice.  In some ways it is actually easier in Eclipse / IntelliJ, which have good integration with gradle, for a local debug.  (I.e.. right-click-run/debug on a gradle task that computes the package ID would save the end-user from digging this out manually from the CLI and loading into the main() scope for the run target.)